core/arch/instrlist.c

```diff
@@ -481,15 +492,18 @@ instrlist_encode_to_copy(dcontext_t *dcontext, instrlist_t *ilist, byte *copy_pc
     instr_t *inst;
     int len = 0;
 #ifdef ARM
-    /* XXX i#1734: reset encdoe state to avoid any stale encode state
+    /* XXX i#1734: reset encode state to avoid any stale encode state
      * or dangling pointer.
      */
     if (instr_get_isa_mode(instrlist_first(ilist)) == DR_ISA_ARM_THUMB)
         encode_reset_it_block(dcontext);
 #endif
-    if (has_instr_jmp_targets || max_pc != NULL) {
+    if (IF_DEBUG_ELSE(true, has_instr_jmp_targets) || max_pc != NULL) {
         /* must set note fields first with offset, or compute length */
         for (inst = instrlist_first(ilist); inst; inst = instr_get_next(inst)) {
+            if (IF_DEBUG_ELSE(!has_instr_jmp_targets, false)
+                    && TEST(INSTR_OPERANDS_VALID, (inst)->flags))
+                DOCHECK(3, { check_inst_opnds(inst); });
```

Comment:
I assume you mean two separate passes in the source code?  At runtime there would only be one, as the DOCHECK walk would only do the walk if !has_instr_jmp_targets.

My argument is that the DOCHECK loop's body shares absolutely nothing with the existing loop body: we don't want to set the note and we don't want to get the length (which sometimes needs a private encode into an allocated buffer that is then copied later, so it's is less efficient then encoding directly to the dest).  So we're duplicating part of the if condition and the for() line is all, to have IMHO less cluttered code.  WDYT?  If you really prefer it all in one loop with extra conditionals, just make sure DEBUG_CHECKS is set to false for release build (doesn't look like it today).